internal/pkg/zeromq/client.go

```diff
@@ -0,0 +1,231 @@
+// Since 0mq pub-sub pattern has different pub socket type and sub socket one
+// the socket initialization and connection are delayed to Publish and Subscribe calls, respectively
+func (client *zeromqClient) Connect() error {
+	return nil
+}
+
+// Publish sends a message via zeromq with the specified topic
+func (client *zeromqClient) Publish(message types.MessageEnvelope, topic string) error {
+
+	msgQueueURL := client.config.PublishHost.GetHostURL()
+	var err error
+
+	if client.publisher == nil {
+
+		if client.publisher, err = zmq.NewSocket(zmq.PUB); err != nil {
+			return err
+		}
+		if conErr := client.publisher.Bind(msgQueueURL); conErr != nil {
+			// wrapping the error with msgQueueURL info:
+			return fmt.Errorf("error: %v [%s]", conErr, msgQueueURL)
+		}
+
+		// allow some time for socket binding before start publishing
+		time.Sleep(300 * time.Millisecond)
+	}
+
+	marshaledMsg, err := json.Marshal(message)
+	if err != nil {
+		return err
+	}
+
+	totalLength, err := client.publisher.SendMessage(topic, marshaledMsg)
+
+	if err != nil {
+		return err
+	} else if totalLength != len(topic)+len(marshaledMsg) {
+		return errors.New("the length of the sent messages does not match the expected length")
+	}
+
+	return err
+}
+
+// Subscribe subscribes to each of the specified topics and posts messages receive to the topic's channel. Any
+// errors encountered while waiting/processing message are sent to the messageErrors channel
+func (client *zeromqClient) Subscribe(topics []types.TopicChannel, messageErrors chan error) error {
+
+	if len(topics) == 0 {
+		return errors.New("no topic(s) specified")
+	} else if len(topics) > maxZeroMqSubscribeTopics {
+		return fmt.Errorf("number of topics(%d) exceeds the maximum capacity(%d)", len(topics), maxZeroMqSubscribeTopics)
+	}
+
+	client.quitSubscribe = make(chan interface{})
+	client.messageErrors = messageErrors
+	client.subscribers = make([]*zeromqSubscriber, len(topics))
+	var errorsSubscribe []error
+	var err error
+
+	for idx, topic := range topics {
+		client.subscribers[idx], err = client.subscribeTopic(&topic)
+		if err != nil {
+			errorsSubscribe = append(errorsSubscribe, err)
+		}
+	}
+
+	if len(errorsSubscribe) == 0 {
+		return nil
+	}
+
+	var errorStr string
+	for _, err := range errorsSubscribe {
+		errorStr = errorStr + fmt.Sprintf("%s  ", err.Error())
+	}
+	return errors.New(errorStr)
+}
+
+func (client *zeromqClient) subscribeTopic(topic *types.TopicChannel) (*zeromqSubscriber, error) {
+
+	subscriber := zeromqSubscriber{}
+	msgQueueURL := client.config.SubscribeHost.GetHostURL()
+	if err := subscriber.init(msgQueueURL, topic); err != nil {
+		return nil, err
+	}
+
+	go func(topic *types.TopicChannel) {
+
+		_ = subscriber.connection.SetSubscribe(topic.Topic)
+		for {
+
+			payloadMsg, err := subscriber.connection.RecvMessage(0)
+
+			select {
+			case <-client.quitSubscribe:
+				_ = subscriber.connection.SetLinger(time.Duration(0))
+				_ = subscriber.connection.Close()
```

Comment:
ignoring close error might lead to incorrect unit test?

Reply:
This is when the subscribe Go func the stopping. Nothing to do with the error.

Reply:
I agree about the subscriber.connection.Close() err I will add that.